.circleci/config.yml

```diff
@@ -75,106 +66,327 @@ commands:
+          environment: #Setting this ENV variable prevents launching a new Docker within build-study-sh script
             USE_DOCKER: false
-      # todo: conditionally don't build if we already have cached the tarball for the SHA+ENV somewhere (circle artifacts?  github package manager?  GCP bucket?
-      # for better build reproducibility, we want to move the same build artifacts to different environments and not rebuild them and
-      # suffer the pain of npm caching, package-lock, and auto-upgrade pitfalls
+      - run:
+          name: Setup gcloud context
+          command: |
+            set -u
+            ./readvault.sh secret/pepper/${ENVIRONMENT}/v1/conf .data.gcp.serviceKey | gcloud auth activate-service-account --key-file=-
+            gcloud --quiet config set compute/zone ${GOOGLE_COMPUTE_ZONE}
+      - run:
+          name: Deploy to GAE
+          command: |
+            set -u
+            gcloud app deploy --version="${SHORT_GIT_SHA}" --stop-previous-version --project "broad-ddp-${ENVIRONMENT}" "${ANGULAR_PROJECT_DIR_PATH}/app.yaml"
+          working_directory: ~/repo/ddp-workspace
+
+  setup-shared-env:
+    description: "Setup shared ENV vars used by multiple scripts"
+    parameters:
+      study_key:
+        type: string
+    steps:
+      - run:
+          name: Set environment variables to build <<parameters.study_key>>
+          command: |
+            echo 'export ANGULAR_PROJECT_NAME=ddp-<<parameters.study_key>>' >> $BASH_ENV
+            echo 'export ANGULAR_PROJECT_DIR_PATH=projects/ddp-<<parameters.study_key>>' >> $BASH_ENV
+            source $BASH_ENV
+            OUTPUT_DIR="${ANGULAR_PROJECT_DIR_PATH}/dist"
+            echo "export OUTPUT_DIR=$OUTPUT_DIR" >> $BASH_ENV
+            echo "export SHORT_GIT_SHA=`echo '<< pipeline.git.revision >>' | cut -c 1-7`" >> $BASH_ENV
+            echo 'export BUILDS_BUCKET=ddp-angular-builds' >> $BASH_ENV
+            source $BASH_ENV
+
+  setup-build-workspace:
+    description: "Setup workspace used by all projects"
+    steps:
+      - checkout:
+          path: ~/repo
       - restore_cache:
           key: *npm_cache_key
       - run:
-          name: <<parameters.study_key>> npm install
+          name: npm install
           command: npm install
           working_directory: ~/repo/ddp-workspace
       - save_cache:
           key: *npm_cache_key
           paths:
             - ./node_modules
+
+  build-app:
+    description: "Build DDP study app"
+    steps:
       - run:
-          name: <<parameters.study_key>> ng build
-          command: ng build $ANGULAR_PROJECT_NAME $NG_BUILD_OPTS --base-href=/ --output-path=$ANGULAR_PROJECT_DIR_PATH/dist --verbose=true --progress=true
+          name: ng build
+          command: |
+            set -u
+            ng build $ANGULAR_PROJECT_NAME --prod --aot --base-href=/ --output-path=$OUTPUT_DIR --verbose=true --progress=true
           working_directory: ~/repo/ddp-workspace
```

Comment:
I just force aot (though my understanding is that the `prod` flag picks up the build settings from Angular config file anyways, so might be redundant.